### DISCUSSION TITLE
Allow easy way to implement cloud shadows in 3D scenes

### DISCUSSION THREAD

**Discussion by Ophiolith** [godotengine/godot-proposals#5871] [3D]
I have been experimenting with different ways to implement cloud shadows into a 3D scene in Godot 4, and frankly, the experience has been an exercise in frustration. Here are some observations:

- Godot's options for projecting textures are limited and do not allow for scrolling textures without some nasty workarounds. Projector output is also extremely grainy when magnified and does not resolve well even on surfaces with high frequency detail. Directional lights do not support projectors, and again, the only workarounds for this are very kludgy.
- Unreal achieves this effect by allowing you to add a 'light function' to your directional light which I presume works as a shader. Append a couple of noise textures to the output of the light and that's it. Godot does not have a comparable feature for DirectionalLight3D.
- Various Blender tutorials suggest using a large plane mesh with a scrolling noise texture with alpha hash transparency, however the effect is extremely noisy and doesn't properly gradiate.
- I've seen a couple of suggestions that writing a SkyShader is the way to go, particularly modifying the radiance cubemap with scrolling perlin noise, however at least in VisualShader there is no output parameter for radiance, only input. This method also inherently precludes using other background types: Color, canvas, camera feed, procedural, physical and panorama skies cannot be used.

I don't really have a solid proposal for how this should be handled, as I don't know enough about the internals of Godot's renderer to suggest a feasible solution. That said, my gut tells me this seems like the sort of broadly useful feature that would fit into WorldEnvironment, and falls within the sort of feature list that environment artists probably shouldn't have to write a custom shader for. I'd welcome any suggestions on how to implement this, and hope this sparks some talk about the topic.

**Comment by Calinou** [comment]
Allowing DirectionalLight3D to have a tiling projector texture drawn in world coordinates (with UV offset and UV scale parameters) is likely the easiest solution. You could then animate the UV offset and/or scale to make the cloud shadow move over time.

In the meantime, you can use a large Decal with a NoiseTexture that spans the whole scene (see https://github.com/Calinou/godotcon-2021-decals). This Decal can be moved and rotated with an AnimationPlayer. The downside is that unless you use the decal Cull Mask property creatively, interior/shaded areas will also receive cloud shadows.

**Reply by Ophiolith** [reply]
Thank you for the feedback! I hadn't considered using decals, but that does seem like a good stop-gap solution. For my particular use-case, having DirectionalLight3D be able to use a tiled texture would absolutely suffice and would be hugely useful. I've only just started to get my feet wet with programming, so I suspect trying to code something like this would be way beyond my skillset, but I'd be eager to at least test and give feedback if there was ever interest?
